dhall/src/Dhall/Core.hs

```diff
@@ -1735,8 +1735,17 @@ normalizeWithM ctx e0 = loop (denote e0)
         case r' of
             RecordLit kvs ->
                 case Dhall.Map.lookup x kvs of
-                    Just v  -> loop v
+                    Just v  -> pure v
                     Nothing -> Field <$> (RecordLit <$> traverse loop kvs) <*> pure x
+            Project r_ _ -> loop (Field r_ x)
+            Prefer l (RecordLit kvs) -> case Dhall.Map.lookup x kvs of
```

Comment:
I think what you want to do here is use a pattern guard, like this:

```haskell
Prefer l (RecordLit kvs) | Just v <- Dhall.Map.lookup x kvs -> pure v
Prefer (RecordLit kvs) r_ | not (Dhall.Map.member x kvs) -> loop (Field r_ x)
…
```

The reason I suggest this is that if you handle the case for `Nothing` here then the normalizer won't attempt other simplifications.

Same comment for `Dhall.Eval`

Reply:
Are you sure? Actually the case for `Nothing` is also an important simplification and calls `loop` on the simplified expression…

Reply:
Oh whoops, I misread this.  You're right